File: sdks/python/copf/storage.py
"""
ConceptStorage interface and in-memory implementation.

Each concept gets its own isolated storage instance, organized by relation.
This matches the TypeScript ConceptStorage interface (Section 6.8).
"""

from __future__ import annotations

import time
from typing import Any, Protocol


class ConceptStorage(Protocol):
    """Storage interface passed to every concept action handler."""

    async def get(self, relation: str, key: str) -> dict[str, Any] | None:
        """Get a single entry by key from a relation."""
        ...

    async def put(self, relation: str, key: str, value: dict[str, Any]) -> None:
        """Store an entry in a relation. Overwrites if key exists."""
        ...

    async def delete(self, relation: str, key: str) -> bool:
        """Delete an entry by key. Returns True if it existed."""
        ...

    async def find(
        self, relation: str, args: dict[str, Any] | None = None
    ) -> list[dict[str, Any]]:
        """Find entries in a relation, optionally filtered by args."""
        ...


class InMemoryStorage:
    """In-memory ConceptStorage implementation for testing and development.

    Data is organized as: relations[relation_name][key] = {value, meta}
    """

    def __init__(self) -> None:
        self._relations: dict[str, dict[str, dict[str, Any]]] = {}

    def _ensure_relation(self, relation: str) -> dict[str, dict[str, Any]]:
        if relation not in self._relations:
            self._relations[relation] = {}
        return self._relations[relation]

    async def get(self, relation: str, key: str) -> dict[str, Any] | None:
        rel = self._ensure_relation(relation)
        entry = rel.get(key)
        return entry["value"] if entry else None

    async def put(self, relation: str, key: str, value: dict[str, Any]) -> None:
        rel = self._ensure_relation(relation)
        rel[key] = {
            "value": value,
            "meta": {"lastWrittenAt": time.time()},
        }

    async def delete(self, relation: str, key: str) -> bool:
        rel = self._ensure_relation(relation)
        if key in rel:
            del rel[key]
            return True
        return False

    async def find(
        self, relation: str, args: dict[str, Any] | None = None
    ) -> list[dict[str, Any]]:
        rel = self._ensure_relation(relation)
        results = [entry["value"] for entry in rel.values()]
        if args:
            results = [
                r
                for r in results
                if all(r.get(k) == v for k, v in args.items())
            ]
        return results

File: sdks/go/copf/storage.go
package copf

import (
	"sync"
	"time"
)

// Storage is the concept storage interface. Each concept gets its own
// isolated storage instance, organized by relation.
// Matches the TypeScript ConceptStorage interface (Section 6.8).
type Storage interface {
	Get(relation, key string) (map[string]any, bool)
	Put(relation, key string, value map[string]any)
	Delete(relation, key string) bool
	Find(relation string, args map[string]any) []map[string]any
}

// InMemoryStorage is a thread-safe in-memory Storage implementation.
type InMemoryStorage struct {
	mu        sync.RWMutex
	relations map[string]map[string]entry
}

type entry struct {
	Value       map[string]any
	LastWritten time.Time
}

// NewInMemoryStorage creates a new empty in-memory storage.
func NewInMemoryStorage() *InMemoryStorage {
	return &InMemoryStorage{
		relations: make(map[string]map[string]entry),
	}
}

func (s *InMemoryStorage) ensureRelation(relation string) map[string]entry {
	if _, ok := s.relations[relation]; !ok {
		s.relations[relation] = make(map[string]entry)
	}
	return s.relations[relation]
}

func (s *InMemoryStorage) Get(relation, key string) (map[string]any, bool) {
	s.mu.RLock()
	defer s.mu.RUnlock()

	rel := s.ensureRelation(relation)
	e, ok := rel[key]
	if !ok {
		return nil, false
	}
	return e.Value, true
}

func (s *InMemoryStorage) Put(relation, key string, value map[string]any) {
	s.mu.Lock()
	defer s.mu.Unlock()

	rel := s.ensureRelation(relation)
	rel[key] = entry{
		Value:       value,
		LastWritten: time.Now(),
	}
}

func (s *InMemoryStorage) Delete(relation, key string) bool {
	s.mu.Lock()
	defer s.mu.Unlock()

	rel := s.ensureRelation(relation)
	if _, ok := rel[key]; ok {
		delete(rel, key)
		return true
	}
	return false
}

func (s *InMemoryStorage) Find(relation string, args map[string]any) []map[string]any {
	s.mu.RLock()
	defer s.mu.RUnlock()

	rel := s.ensureRelation(relation)
	var results []map[string]any

	for _, e := range rel {
		if matchesArgs(e.Value, args) {
			results = append(results, e.Value)
		}
	}
	return results
}

func matchesArgs(value, args map[string]any) bool {
	if args == nil {
		return true
	}
	for k, v := range args {
		if value[k] != v {
			return false
		}
	}
	return true
}
